Add `Leaf` with `start` and `end` props

zbeyens opened the pull request:
**Description**
Add optional `position` property to leaves returned by `Text.decorations()`. This makes it easier to track the exact position and boundary of each leaf within its parent text node, especially useful when you need to render elements relative to text nodes that have been split by decorations.

**Example**
Before:
```js
Text.decorations({ text: 'abc', mark: 'bold' }, [...])
// Returns:
[
  { text: 'a', mark: 'bold' },
  { text: 'b', mark: 'bold', decoration: 'highlight' },
  { text: 'c', mark: 'bold' }
]
```

After:
```js
Text.decorations({ text: 'abc', mark: 'bold' }, [...])
// Returns leaves with position info only when split by decorations:
[
  { 
    text: 'a', 
    mark: 'bold',
    position: { start: 0, end: 1, isFirst: true, isLast: false }
  },
  { 
    text: 'b', 
    mark: 'bold', 
    decoration: 'highlight',
    position: { start: 1, end: 2, isFirst: false, isLast: false }
  },
  { 
    text: 'c', 
    mark: 'bold',
    position: { start: 2, end: 3, isFirst: false, isLast: true }
  }
]

// Single leaf = no position info needed:
[{ text: 'abc', mark: 'bold' }]
```

**Context**
When working with decorated text nodes, it's often necessary to know the exact position of each leaf within the original text. A common use case is rendering an element exactly once after a text node, even when that text node has been split into multiple leaves by decorations.

For example, if you want to render a tooltip after a text node, using `renderLeaf` directly would cause the tooltip to render multiple times (once per leaf) because `renderLeaf` is called for each decorated segment. With the new `position` property, you can now reliably check:

```jsx
const renderLeaf = ({ leaf, text, ...props }) => {
  return (
    <>
      <span {...props.attributes}>{props.children}</span>
      {/* Render tooltip only once, after the last leaf of this text node */}
      {leaf.position?.isLast && <Tooltip />}
    </>
  )
}
```

The `position` property is only added when a text node is actually split by decorations, keeping the leaf objects clean and simple in the common case where no splitting occurs. Potential naming conflicts with user-defined `position` mark could be resolved by referring to the `Text` node instead.

12joan commented:
I wonder if it would be better to put this information in a separate property of `RenderLeafProps` from `leaf`? That would prevent unintuitive behaviour like `Node.extractProps` returning `position` for decorated text nodes when called from `renderLeaf`, and it would keep the interface of `leaf` consistent in all cases.

The property could be named to make it clearer that it refers to the position of the leaf within a decoration, rather than its position within the parent element, for example.

zbeyens commented:
Good point @12joan, I've refactored to use `leafPosition` instead. I've also added `renderText` that would be an alternative solution to our use-case.

12joan commented on packages/slate-react/src/components/text.tsx:
```diff
@@ -65,17 +78,32 @@ const Text = (props: {
+      text,
+      children: textContent,
+      attributes: {
+        'data-slate-node': 'text',
+        ref: callbackRef,
```
We probably don't want to apply `data-slate-node="text"` to multiple different layers of the DOM hierarchy, and we certainly don't want to apply the ref twice. I would probably omit `attributes` from `renderText`, or possibly leave it as an empty object if we want to preserve the ability to add a distinct `ref` later without it being a breaking change. I can't think of any reason why that would be necessary, though.

zbeyens replied:
I took those attributes from the default behavior for backward compatibility.

Unless I'm missing something, there can't be nested renderText.

12joan replied:
`textContent` also has those attributes, and is passed as `children` to `renderText`.

```tsx
  const textContent = (
    <span data-slate-node="text" ref={callbackRef}>
      {children}
    </span>
  )

  if (renderText) {
    return renderText({
      text,
      children: textContent,
      attributes: {
        'data-slate-node': 'text',
        ref: callbackRef,
      },
    })
  }
  ```

If `renderText` is implemented as `({ attributes, children }) => <span {...attributes}>{children}</span>`, then the resulting JSX is:

```tsx
<span data-slate-node="text" ref={refCallback}>
  <span data-slate-node="text" ref={refCallback}>
    {children}
  </span>
</span>
```

zbeyens replied:
Good catch, I meant to use children instead of textContent there. Updating...

12joan replied:
Got it. Since that's your intent, perhaps we could use the same pattern as `renderElement` to provide a default `renderText` that just renders a span with those attributes? That would eliminate the need for an if statement here.

https://github.com/ianstormtaylor/slate/blob/main/packages/slate-react/src/components/element.tsx#L44

zbeyens replied:
Done

12joan commented:
> Good point @12joan, I've refactored to use `leafPosition` instead. I've also added `renderText` that would be an alternative solution to our use-case.

Neat solution with `renderText`!
